imjoy/connection/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""Provide the socketio connection manager."""
+import asyncio
+from functools import partial
+
+import socketio
+from aiohttp import web
+
+from imjoy.const import ENG
+from .decorator import partial_coro
+from .handler import register_services
+from .server import setup_app, run_app
+
+
+def create_connection_manager(eng):
+    """Create a socketio connection and return the connection instance."""
+    # An event handler can be found like this:
+    # handler = sio.handlers[namespace][event]
+    # ALLOWED_ORIGINS = [opt.base_url, 'http://imjoy.io', 'https://imjoy.io']
+    sio = socketio.AsyncServer()
+    app = web.Application()
+    app[ENG] = eng
+    sio.attach(app)
+    setup_app(eng, app)
+    return ConnectionManager(eng, app, sio)
+
+
+def register_event_handler(eng, event, handler=None, namespace=None):
+    """Register a socketio event handler."""
+    # pylint: disable=protected-access
+    if handler is None:
+        handler = event
+        event = handler._ws_event
+        namespace = handler._ws_namespace
+    if asyncio.iscoroutinefunction(handler):
+        injected_handler = partial_coro(handler, eng)
+    else:
+        injected_handler = partial(handler, eng)
+    eng.conn.sio.on(event, handler=injected_handler, namespace=namespace)
+
+
+class ConnectionManager:
+    """Represent a connection manager for socketio event handler and session data."""
+
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, eng, app, sio):
+        """Set up connection instance attributes."""
+        self.eng = eng
+        self.app = app
+        self.sio = sio
+        self.data = ConnectionData()
+
+    def start(self):
+        """Start the connection."""
+        self._register_services()
+        run_app(self.eng, self.app)
+
+    def register_event_handler(self, event, handler=None, namespace=None):
+        """Register a socketio event handler."""
+        register_event_handler(self.eng, event, handler=handler, namespace=namespace)
+
+    def _register_services(self):
+        """Register event handlers for internal services."""
+        register_services(self.eng, register_event_handler)
+
+
+class ConnectionData:
```

Comment:
I think we should get rid of this layer of wrapping, place all the data directly inside `ConnectionManager`.
Too many layers of wrapping make the code a bit tedious, also decrease performance to some extent.

For example, in `plugin.py`, we have many like this: `eng.conn.data.plugin_sessions`, and this should be `eng.conn.plugin_sessions`. or even `eng.plugin_sessions`.

Reply:
If the data belongs to the connection it should be put on the connection. But we can make a shortcut on the engine instance and copy the data reference to an attribute on the engine too.
```py
def __init__(self, opt):
    self.conn = None
    self.data = None

def setup(self):
    ...
    self.data = self.conn.data
```

Reply:
We can also replace the data class with a `dotdict` class, which we already have in the worker utils. That will allow easier extension.

Reply:
I just implemented this shortcut and dotdict.

And also renamed `data` to `store`, to make it more like the popular state management `store pattern`, e.g. in Vuejs (https://vuejs.org/v2/guide/state-management.html)